### PR TITLE
feat(lightwalletd): add build and CI pipeline

### PR DESCRIPTION
**Pull request by gustavovalverde**
## Motivation

This is the CI for the unit test made available with https://github.com/ZcashFoundation/zebra/pull/3619

## Solution

- Rewrite lightwalletd Dockerfile to build a lightweight binary
- Use lightwalletd image binary inside zebra's image
- Create a step to run `lightwalletd_integration` tests

# Extras:
- Validate any broken changes made to lightwalletd by building on our side
- Validate any broken code in lightwalletd with linter

## Review
@teor2345 and @dconnolly can review this

### Reviewer Checklist

  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work
- Make a reusable action, so we don't trigger multiple workflows with the same image.

### CODE REVIEW

**Comment by gustavovalverde** [comment]
@teor2345 @oxarbitrage for some reason the tests is not being selected. Please confirm the command being used is the correct one: https://github.com/ZcashFoundation/zebra/pull/3657/files#diff-faff1af3d8ff408964a57b2e475f69a6b7c7b71c9978cccc8f471798caac2c88R165

File: .github/workflows/zcash-lightwalletd.yml [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        context: .
+        file: ./zebra/docker/Dockerfile.lightwalletd
+        tags: |
+          ${{ env.GAR_BASE }}/${{ env.IMAGE_NAME }}:latest
+          ${{ env.GAR_BASE }}/${{ env.IMAGE_NAME }}:${{ env.GITHUB_SHA_SHORT }}
+        push: true
+        cache-from: type=gha
+        cache-to: type=gha,mode=max
+
+  lint-lightwalletd:
```
Comment (teor2345) [inline .github/workflows/zcash-lightwalletd.yml]: Scope Creep: We didn't write the lightwalletd code, and we're not planning on changing it.

Is there any benefit to linting it?

I'm concerned that linter errors might make test debugging really confusing.

File: docker/Dockerfile.lightwalletd [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+ARG LWD_HTTP_PORT
+
+WORKDIR /
+
+COPY --from=build /lightwalletd /usr/local/bin
+COPY --from=build $ZCASHD_CONF_PATH ./
+
+EXPOSE 9067
+EXPOSE 9068
+EXPOSE 8232
```
Comment (teor2345) [inline docker/Dockerfile.lightwalletd]: Security: 8232 is the Zebra or zcashd RPC port, so we don't want it exposed to the public internet.

(And there is nothing listening on this port in this image.)

File: docker/Dockerfile.lightwalletd [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+CMD ["--no-tls-very-insecure", "--grpc-bind-addr=0.0.0.0:9067",  "--http-bind-addr=0.0.0.0:9068", "--log-file=/dev/stdout", "--log-level=7"]
+
+##
+## Deploy
+##
+FROM debian:bullseye-slim as runtime
+
+ARG ZCASHD_CONF_PATH
+ARG LWD_GRPC_PORT
+ARG LWD_HTTP_PORT
```
Comment (teor2345) [inline docker/Dockerfile.lightwalletd]: Question: the port arguments aren't actually used for anything, should they be deleted? Or should they replace the fixed port values?

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -146,6 +146,26 @@ jobs:
...
+        with:
+          persist-credentials: false
+
+      - name: Inject slug/short variables
+        uses: rlespinasse/github-slug-action@v4
+
+      - name: Run tests with included lightwalletd binary
+        run: |
+          docker pull ${{ env.GAR_BASE }}/zebrad-test:${{ env.GITHUB_EVENT_PULL_REQUEST_HEAD_SHA_SHORT || env.GITHUB_SHA_SHORT }}
+          docker run -e ZEBRA_SKIP_IPV6_TESTS -e ZEBRA_TEST_LIGHTWALLETD --name zebrad-tests -t ${{ env.GAR_BASE }}/zebrad-test:${{ env.GITHUB_EVENT_PULL_REQUEST_HEAD_SHA_SHORT || env.GITHUB_SHA_SHORT }} cargo test --locked --release --features enable-sentry --test acceptance lightwalletd_integration -- --ignored
```
Comment (teor2345) [inline .github/workflows/test.yml]: Bug: The test name filter comes after `--`.

1. There are multiple instances of this bug in CI, can you please fix the rest in another high-priority PR?
2. I'd like to fail CI test jobs when no tests are actually run. Can you please open a high-priority ticket for that?

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -146,6 +146,26 @@ jobs:
           docker pull ${{ env.GAR_BASE }}/${{ env.IMAGE_NAME }}:${{ env.GITHUB_EVENT_PULL_REQUEST_HEAD_SHA_SHORT || env.GITHUB_SHA_SHORT }}
           docker run -e ZEBRA_SKIP_IPV6_TESTS --name zebrad-tests -t ${{ env.GAR_BASE }}/${{ env.IMAGE_NAME }}:${{ env.GITHUB_EVENT_PULL_REQUEST_HEAD_SHA_SHORT || env.GITHUB_SHA_SHORT }} cargo test --locked --release --features enable-sentry --test acceptance sync_large_checkpoints_ -- --ignored
```
Comment (teor2345) [inline .github/workflows/test.yml:147]: The same test filter bug is also present in this job, which means that these tests haven't run for weeks.

**Review by teor2345 — DISMISSED** [review]
> @teor2345 @oxarbitrage for some reason the tests is not being selected. Please confirm the command being used is the correct one: https://github.com/ZcashFoundation/zebra/pull/3657/files#diff-faff1af3d8ff408964a57b2e475f69a6b7c7b71c9978cccc8f471798caac2c88R165

The test filter name comes after `--`.

It looks like we also missed this in our reviews of some earlier refactors.

File: .github/workflows/zcash-lightwalletd.yml (hunk 1 above)
Comment (gustavovalverde) [reply .github/workflows/zcash-lightwalletd.yml]: There's no real benefit. If this feels more noisy than valuable, I'll remove it.

File: docker/Dockerfile.lightwalletd (hunk 3 above)
Comment (gustavovalverde) [reply docker/Dockerfile.lightwalletd]: I'm not using those, but I kept those values to maintain backward compatibility with their docker-compose and arguments they're using. https://github.com/adityapk00/lightwalletd/blob/master/docker-compose.yml#L11

Comment (gustavovalverde) [reply docker/Dockerfile.lightwalletd:49]: Fixed at [0584246](https://github.com/ZcashFoundation/zebra/pull/3657/commits/0584246328ce5ea70ba5b11bc9b6aa6c1e622842)

File: .github/workflows/test.yml (hunk 4 above)
Comment (gustavovalverde) [reply .github/workflows/test.yml]: Fixed at [668481a](https://github.com/ZcashFoundation/zebra/pull/3657/commits/668481ad49f3e90a7782df6472d1af5e6c7a2c5d)

File: docker/Dockerfile.lightwalletd (hunk 2 above)
Comment (gustavovalverde) [reply docker/Dockerfile.lightwalletd]: Fixed at [0584246](https://github.com/ZcashFoundation/zebra/pull/3657/commits/0584246328ce5ea70ba5b11bc9b6aa6c1e622842)

File: .github/workflows/zcash-lightwalletd.yml (hunk 1 above)
Comment (gustavovalverde) [reply .github/workflows/zcash-lightwalletd.yml]: Removed with [d37923c](https://github.com/ZcashFoundation/zebra/pull/3657/commits/d37923c638a5b5ab82d132c77f383182b2de2f8c)

File: .github/workflows/test.yml (hunk 5 above)
Comment (gustavovalverde) [reply .github/workflows/test.yml:147]: This is odd, as we've even seen the logs of this tests running.

This is from a recent run: https://github.com/ZcashFoundation/zebra/runs/5345011546?check_suite_focus=true#step:4:127

**Comment by gustavovalverde** [comment]
@teor2345 after applying https://github.com/ZcashFoundation/zebra/pull/3657#discussion_r815274026, we have the same issue https://github.com/ZcashFoundation/zebra/runs/5353929503?check_suite_focus=true#step:4:133

**Comment by conradoplg** [comment]
> @teor2345 after applying [#3657 (comment)](https://github.com/ZcashFoundation/zebra/pull/3657#discussion_r815274026), we have the same issue https://github.com/ZcashFoundation/zebra/runs/5353929503?check_suite_focus=true#step:4:133

`--ignored` runs _only_ ignored tests, and `lightwalletd_integration` isn't one of them. So it ends up not running anything. You'll need to remove `--ignored`

File: .github/workflows/test.yml (hunk 5 above)
Comment (conradoplg) [reply .github/workflows/test.yml:147]: You're right, I tested locally and it works even before `--`. The [documentation](https://doc.rust-lang.org/cargo/commands/cargo-test.html#examples) is confusing and mentions both.

File: docker/Dockerfile.lightwalletd (hunk 3 above)
Comment (teor2345) [reply docker/Dockerfile.lightwalletd]: Can you add a comment documenting why they are there?

File: .github/workflows/test.yml (hunk 5 above)
Comment (teor2345) [reply .github/workflows/test.yml:147]: Wow, that's really unhelpful documentation.

(I'll go find the test that isn't running, I thought it was this one.)

Comment (teor2345) [reply .github/workflows/test.yml:147]: I opened this ticket for the test that isn't running:
- #3673

File: .github/workflows/test.yml (hunk 4 above)
Comment (teor2345) [reply .github/workflows/test.yml]: I opened or found these tickets for follow up:
- #3673 
- #2653

**Review by teor2345 — DISMISSED** [review]
Looks good, let's merge.

(The extra comment isn't a blocker.)

**Comment by gustavovalverde** [comment]
Just to be sure here @teor2345. 

Based on @conradoplg comment https://github.com/ZcashFoundation/zebra/pull/3657#issuecomment-1054303368. Are we removing the `--ignored` flag from the command, or are you adding this test to the ignored ones?

**Comment by teor2345** [comment]
> Just to be sure here @teor2345.
> 
> Based on @conradoplg comment [#3657 (comment)](https://github.com/ZcashFoundation/zebra/pull/3657#issuecomment-1054303368). Are we removing the `--ignored` flag from the command, or are you adding this test to the ignored ones?

Please remove the `--ignored` flag in the CI job.

(There's no point having both `#[ignored]` and an environmental variable. That's just confusing!)

File: docker/Dockerfile.lightwalletd (hunk 3 above)
Comment (gustavovalverde) [reply docker/Dockerfile.lightwalletd]: Fixed by:[41829d4](https://github.com/ZcashFoundation/zebra/pull/3657/commits/41829d4722e566b32bbe0b6c679d4c196f7904a8)

**Comment by gustavovalverde** [comment]
@teor2345 is the last run fail something expected?

**Comment by teor2345** [comment]
> @teor2345 is the last run fail something expected?

We added another RPC, so the error mentions the next RPC now. I'll edit this PR with the fix.

**Comment by teor2345** [comment]
> @teor2345 is the last run fail something expected?

Actually, it turns out that zcash/lightwalletd and adityapk00/lightwalletd behave slightly differently. I've updated the tests to match adityapk00/lightwalletd.

Once PR #3628 merges, it will contain the fix in commit de9069f23909588cf0078c5a224ee42144401867.

File: docker/zcash-lightwalletd/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+ARG ZCASHD_CONF_PATH
+
+RUN set -ex; \
+  { \
+    echo "rpcuser=zcashrpc"; \
+    echo "rpcpassword=`head /dev/urandom | tr -dc A-Za-z0-9 | head -c 13 ; echo ''`" \
+    echo "rpcbind=127.0.0.1"; \
+    echo "rpcport=8232"; \
+  } > "${ZCASHD_CONF_PATH}"
```
Comment (dconnolly) [inline docker/zcash-lightwalletd/Dockerfile:37]: Heh

File: docker/zcash-lightwalletd/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+# Do not use `go get` as it updates the requirements listed in your go.mod file.
+# `go mod download` does not add new requirements or update existing requirements.
+RUN go mod download
+
+# Copy local code to the container image.
+COPY . ./
+
+# Build and install the binary.
+RUN go build -v -o /lightwalletd
```
Comment (dconnolly) [inline docker/zcash-lightwalletd/Dockerfile:27]: :gopher:

File: docker/zcash-lightwalletd/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+WORKDIR /
+
+COPY --from=build /lightwalletd /usr/local/bin
+COPY --from=build $ZCASHD_CONF_PATH ./
+
+EXPOSE 9067
+EXPOSE 9068
+
+ENTRYPOINT ["lightwalletd"]
+CMD ["--no-tls-very-insecure", "--grpc-bind-addr=0.0.0.0:9067",  "--http-bind-addr=0.0.0.0:9068", "--log-file=/dev/stdout", "--log-level=7"]
```
Comment (dconnolly) [inline docker/zcash-lightwalletd/Dockerfile:61]: 🥰

**Review by dconnolly — APPROVED** [review]

**Comment by dconnolly** [comment]
> > @teor2345 is the last run fail something expected?
> 
> Actually, it turns out that zcash/lightwalletd and adityapk00/lightwalletd behave slightly differently. I've updated the tests to match adityapk00/lightwalletd.
> 
> Once PR #3628 merges, it will contain the fix in commit [de9069f](https://github.com/ZcashFoundation/zebra/commit/de9069f23909588cf0078c5a224ee42144401867).

This has now merged

File: docker/Dockerfile [hunk 9]
```diff
@@ -56,6 +56,7 @@ FROM builder AS tester
 # Pre-download Zcash Sprout and Sapling parameters
 # TODO: do not hardcode the user /root/ even though is a safe assumption
 COPY --from=us-docker.pkg.dev/zealous-zebra/zebra/zcash-params /root/.zcash-params /root/.zcash-params
+COPY --from=us-docker.pkg.dev/zealous-zebra/zebra/lightwalletd /lightwalletd /usr/local/bin
```
Comment (dconnolly) [inline docker/Dockerfile:59]: Easy :D
